### DISCUSSION TITLE
In some scenario you need to use WithPerceptionTracking inside the block.

### DISCUSSION THREAD

**Discussion by alfredcc** [pointfreeco/swift-perception#21] [Q&A] [answered]
### Description

``` swift
import Perception
import SwiftUI

@Perceptible
class ColorModel {
    var color: Color = .clear
}

struct ContentView: View {
    
    @State var model = ColorModel()
    
    var body: some View {
        let _ = print("\(Self.self): tracked change.")
        WithPerceptionTracking {
            VStack {
                ForEach(0 ... 3, id: \.self) { _ in
//                    WithPerceptionTracking {
                        model.color
//                    }
                }
            }
        }
    }
}


#Preview {
    ContentView()
}
```

If I comment out the above content, I will get the following runtime warning:
`Perceptible state was accessed but is not being tracked. Track changes to state by wrapping your view in a 'WithPerceptionTracking' view.`

**Comment by alfredcc** [comment]
In a similar scenario where you are using `GeometryReader`, you need to use `WithPerceptionTracking` inside it. 

```swift
// This indicates that you should use `WithPerceptionTracking` inside the `GeometryReader` block. 
GeometryReader { proxy in
    WithPerceptionTracking {
        // Your code here
    }
}
```

**Comment by stephencelis** [comment, marked as answer]
@alfredcc This is to be expected. `WithPerceptionTracking` is a synchronous call, and so if you hit an escaping closure that is executed later, it must also be instrumented with `WithPerceptionTracking`. If someone figures out a way around this we'd love to see it, but we don't think it's possible to do from outside SwiftUI.

**Reply by PatrickDanino** [reply]
How would you safely perform an asynchronous call, say using a Task?

For example, this simple change using your `Example` causes a runtime warning:

```
WithPerceptionTracking {
  Form {
     ...
  }
  .onAppear {
     Task { model.incrementButtonTapped() }
  }
}
```

**Reply by mbrandonw** [reply]
That is unfortunately just an erroneous warning. [The logic](https://github.com/pointfreeco/swift-perception/blob/42240120b2a8797595433288ab4118f8042214c3/Sources/Perception/PerceptionRegistrar.swift#L209-L215) that shows the error is not accounting for that situation correctly. We will need to look into it more, or if anyone else has ideas please do open a PR or discuss here.

For now you can just ignore that warning.

**Reply by PatrickDanino** [reply]
Thanks for confirming!

**Reply by tonyarnold** [reply]
It'd be wonderful if there were a curated list of places where the tracking needs to be re-instated.

**Reply by mbrandonw** [reply]
Hi @tonyarnold, the [docs](https://swiftpackageindex.com/pointfreeco/swift-perception/main/documentation/perception/withperceptiontracking) do give the definitive guide to where you should install `WithPerceptionTracking`:

> You will also need to use [WithPerceptionTracking](https://swiftpackageindex.com/pointfreeco/swift-perception/main/documentation/perception/withperceptiontracking) in any escaping, trailing closures used in SwiftUI’s various navigation APIs, such as the sheet modifier:

So a curated list would only be able to include the escaping closures that come with SwiftUI, but wouldn't be able to describe ones that come from 3rd party code. 

However, if those docs are unclear or need to be expanded we can certainly do that. Please let us know if you have any feedback on that.

**Reply by tonyarnold** [reply]
That's completely reasonable, @mbrandonw, thanks — I was being lazy and didn't want to have to check the method signatures and behavior of the places in SwiftUI I suspected might be a problem. I appreciate that it's a bit more involved than "here's the itemized list of places in SwiftUI that you'll need to worry about this".

**Reply by m-yeroshenko** [reply]
Hi @mbrandonw !
Could you please attach a working link to the documentation you've mentioned above?
That one is not working.
Thank you!
